File: Categories/Stack/20_validParentheses.py
from collections import defaultdict
from typing import *

class Solution:
    def isValid(self, s: str) -> bool:
        rcd = defaultdict(str, {
            '}' : '{',
            ']' : '[',
            ')' : '(',
        })

        stack = []
        for ch in s:
            if ch in ['{', '[', '(']:
                stack.append(ch)
            else:
                if not stack or stack[-1] != rcd[ch]:
                    return False
                stack.pop()

        return not stack

File: Categories/Stack/20_validParentheses.cs
public class Solution {
    public bool IsValid(string s) {
        var rcd = new Dictionary<char, char> {
            [')'] = '(',
            ['}'] = '{',
            [']'] = '[',
        };

        var stack = new Stack<char>();

        foreach (char ch in s) {
            if (
                ch == '(' || 
                ch == '[' ||
                ch == '{'
            ) {
                stack.Push(ch);
            }
            else {
                if (stack.Count == 0) {
                    return false;
                }

                if (stack.Peek() != rcd[ch]) {
                    return false;
                } else {
                    stack.Pop();
                }
            }
        }

        return stack.Count == 0;
    }
}